### ISSUE TITLE
The final solution for country leader board

### ISSUE THREAD

**Issue by dadiorchen** [Greenstand/treetracker-web-map-client#724]
We need to find a formal solution for the leader board of countries, currently, we use SQL directly on query API to fetch the board, but, the problem is that it is too slow, we need to find a way to offer acceptable performance for querying the board data.

**Comment by dadiorchen** [comment]
Some idea:

* Use airflow to pre-generate the data.
* Use pgpool to pre-cache the SQL

**Comment by dadiorchen** [comment]
Maybe we can combine airflow + query-api services:

1. query-api open API for fetch the data.
2. airflow fetch the API and dump data the db. (question: who do the insertion? airflow or query-api)
3. fetch another API on query-api to get the pre-genrated data.

In this way, airflow job is simpler.

@dagmawig could you give some opinion here?

**Comment by dadiorchen** [comment]
Another thing, we need to consider the leader board for continents.

**Comment by dadiorchen** [comment]
Another solution: utilize the next.js cache system through the contribution by @okdv !

@okdv can you give any input on this?

**Comment by dagmawig** [comment]
Can we do some kind of indexing like what we are trying to accomplish with the search function? Can Solr be used on this?

**Comment by okdv** [comment]
next cache is a possibility here. that being said, it would still require clients load the full request from time to time so it can generate the static pages, so it would probably be a good idea to find a way to expedite that query when it does have to be used still, so its probably better in combo with something as opposed to standalone fix. 

I think an indexer like @dagmawig said could work well, especially if its already going to be indexing the same data for the search function. If you combined it with next caching you would probably have a pretty quick experience i think, im happy to help with benchmarking if we go that route

**Comment by dadiorchen** [comment]
@okdv @dagmawig yes, I think these approaches all make sense, now I tend to use next.js cache as the first step, because this is the most buy for us, it's simple, fast, and speed up the whole site's performance, with the mechanism of pre-cache and enable-it-at-next-short, I think the respond of the site would be really fast.

As to the first visit for a new start of the site, I think we can use airflow to do auto-fetch periodically.

@okdv can you explain more about `im happy to help with benchmarking if we go that route` ? Some pressure test?

**Comment by okdv** [comment]
exactly yea, if we implement some other tools that could be used, e.g. next cache vs airflow vs solr vs combination of them, im happy to help with testing them so we can get some time comparisons down, decide whats worth keeping long term. 

that being said, i think using only next cache can be a solid short term solution. load times when the cache expires might suck but hopefully most sessions wont have to deal with it and we can improve on it later

**Comment by dadiorchen** [comment]
I really wanted a test for the performance, I even consider greenstand needs a test team for this, to pressure test, available test for all our service, if you can push the test for the issue we are discussing, I think that would be a good start. Do you have any plan on this? tools? the resource you need and so on? I will definitely support you on this.

**Comment by dadiorchen** [comment]
Just a month ago, one of our service crashed because of the pressure, the DB can not bear the concurrent requests emitted, we really need some solution for this kind of case.

**Comment by okdv** [comment]
For this bit in particular, I was thinking some pretty basic performance testing, mostly with in built performance api in the browser, maybe use cypress to programmatically test it depending on how everything is already set up. 

That being said, for big picture, I would certainly agree that having some testing or QA team or people in place would be a good idea. stress testing a db or web server can be a little involved, but is certainly doable with stuff like jmeter, depending on our backend, lots of options. 

I also see we already use kibana and grafana in some ways, so simple gathering data about events like that crash and analyzing postmortem is also very valuable. 

i'll see if i cant dig around the backend side of things and find out more about whats already in place and if there is anywhere i can be of some help.

as for this issue, i believe next cache is already implemented on this page (/top?) at the moment, but correct me if im wrong

**Comment by dadiorchen** [comment]
@okdv we used k6 as the tool to do pressure test for our wallet-api before, by a volunteer, https://k6.io/ but JMeter is kind of more battle proved one.

I ran into problem with some test running on github action, so I temporarily close the cache, I plan to recover it and add a switcher to turn it on/off, so 1. we can disable it on CI to avoid unexpected result. 2. Bring convenience when we want to test the effect of the cache.
